compiler/crates/relay-lsp/src/goto_definition/mod.rs

```diff
@@ -120,22 +124,48 @@ fn locate_type_definition(
     extra_data_provider: &dyn LSPExtraDataProvider,
     project_name: StringKey,
     type_name: StringKey,
+    schema: &Arc<SDLSchema>,
+    root_dir: &std::path::Path,
 ) -> Result<GotoDefinitionResponse, LSPRuntimeError> {
+    let type_ = schema.get_type(type_name);
+
+    let schema_location = type_.and_then(|type_| {
+        match type_ {
+            // Why don't these other types have locations?
+            Type::InputObject(_) => None,
+            Type::Enum(_) => None,
+            Type::Interface(_) => None,
+            Type::Scalar(_) => None,
+            Type::Union(_) => None,
+            Type::Object(object_id) => Some(schema.object(object_id).name.location),
+        }
+    });
```

Comment:
If the external data provider as a valid response, we will not end up using this. Would it be possible to mimic what we do in `locate_field_definition` and try to use the provided response if one is given, and then fallback to finding the value in the schema?

Reply:
@captbaritone I tried doing what you said. the nested options / results got pretty hairy. is what I did ok? is there a more idiomatic way of going from `Option<Result<T, E>>` => `Result<T, E>`?